### PR TITLE
Support Patches within Versions

### PR DESCRIPTION
**Pull request by canadaduane**
This adds support for the "Patches: N" header within a version, which implies 3 tiers of headers: HTTP headers, Version Headers, and Patch headers.

### CODE REVIEW

File: server/src/stream.ts [hunk 1]
```diff
@@ -72,11 +72,11 @@ function sendInitialValOnly(res: ServerResponse, opts: StateServerOpts): void {
     ...opts.httpHeaders,
   }
 
-  if (opts.contentType) httpHeaders['content-type'] = opts.contentType
-  if (opts.initialVerson) httpHeaders['version'] = opts.initialVerson
+  if (opts.contentType) httpHeaders['Content-Type'] = opts.contentType
+  if (opts.initialVerson) httpHeaders['Version'] = opts.initialVerson
```
Comment (josephg) [inline server/src/stream.ts]: I'd rather we keep headers lowercased in the code. This is the standard for h2 & going forward, because http headers are case insensitive anyway and being consistent is better for gzip encoding & gzip dictionaries.

File: server/src/stream.ts [hunk 2]
```diff
@@ -72,11 +72,11 @@ function sendInitialValOnly(res: ServerResponse, opts: StateServerOpts): void {
...
 
   let bufData = toBuf(opts.initialValue)
-  httpHeaders['content-length'] = bufData.length
+  httpHeaders['Content-Length'] = bufData.length
```
Comment (josephg) [inline server/src/stream.ts]: and here

File: server/src/stream.ts [hunk 3]
```diff
@@ -99,14 +99,14 @@ export function stream(
   }
 
   const httpHeaders: Record<string, string> = {
-    'cache-control': 'no-cache',
-    'connection': 'keep-alive',
+    'Cache-Control': 'no-cache',
+    'Connection': 'keep-alive',
```
Comment (josephg) [inline server/src/stream.ts]: and here

File: server/src/stream.ts [hunk 4]
```diff
@@ -99,14 +99,14 @@ export function stream(
...
     ...opts.httpHeaders,
   }
 
   let contentType = opts.contentType ?? null
-  if (contentType != null) httpHeaders['content-type'] = contentType
-  if (opts.patchType) httpHeaders['patch-type'] = opts.patchType
+  if (contentType != null) httpHeaders['Content-Type'] = contentType
+  if (opts.patchType) httpHeaders['Patch-Type'] = opts.patchType
```
Comment (josephg) [inline server/src/stream.ts]: and here

File: server/src/stream.ts [hunk 5]
```diff
@@ -144,24 +144,39 @@ export function stream(
...
-        }
-        if (val.version) patchHeaders['version'] = val.version
+        // Whether we have patches or just a version, include the version here
+        if (val.version) versionHeaders['Version'] = val.version
+
+        // Testing alternative braid protocol ideas here:
+        if (val.patchId) versionHeaders['Patch-Id'] = val.patchId
         if (val.patchType && val.patchType !== lastPatchType) {
-          patchHeaders['patch-type'] = lastPatchType = val.patchType
+          versionHeaders['Patch-Type'] = lastPatchType = val.patchType
```
Comment (josephg) [inline server/src/stream.ts]: ^-- lowercase all these headers

File: server/src/stream.ts [hunk 6]
```diff
@@ -144,24 +144,39 @@ export function stream(
 
       for await (const val of stream.iter) {
         if (!connected) break
-        // console.log('got val', val)
 
-        const data = toBuf(val.data)
+        let versionHeaders: Record<string, string> = { ...val.headers }
```
Comment (josephg) [inline server/src/stream.ts]: Is there a standard name for this container object in the spec? My preference is a name like `transactionHeaders` or something. Calling a transaction a 'version' feels like overloading a term thats pretty overloaded already, but I'd like to be consistent about it.

File: server/src/stream.ts [hunk 7]
```diff
@@ -144,24 +144,39 @@ export function stream(
...
 
-        writeHeaders(res, patchHeaders)
-        res.write(data)
+        if (val.patches && val.patches.length > 0) {
+          versionHeaders['Patches'] = `${val.patches.length}`
+          writeHeaders(res, versionHeaders)
+
+          for (let { range, data } of val.patches) {
+            const patchHeaders: Record<string, string> = {
+              'Content-Length': `${data.length}`,
```
Comment (josephg) [inline server/src/stream.ts]: ` data: StringLike`

So data.length isn't right here. We want the byte length, and if data is a string then string length might not be the same as the byte length. Convert data to a buffer and use .length there. (Or some similar approach).

Example 2 should be able to reproduce this bug.

File: examples/1/client.js [hunk 8]
```diff
@@ -1,9 +1,8 @@
-const {subscribe} = require('@josephg/braid-client')
+const { subscribe } = require('@josephg/braid-client')
 
 ;(async () => {
-  const {initialValue, stream} = await subscribe('http://localhost:2001/time')
-  console.log('initial value', initialValue)
-  for await (const {value} of stream) {
+  const { stream } = await subscribe('http://localhost:2001/time')
+  for await (const { value } of stream) {
```
Comment (josephg) [inline examples/1/client.js:5]: Why did you remove initialValue here? Unless I'm missing something, this will now miss out on the first sent value in the stream. (And we won't get any values until after the first patch is sent).

File: client/client.ts [hunk 9]
```diff
@@ -44,67 +45,141 @@ const searchHeaderGap = (buf: Uint8Array): null | [string, number] => {
...
-  // Tiny state machine. We swap back and forth from reading the headers <->
-  // reading data. Every chunk must contain a content-length field.
+async function* readHTTPChunks(res: Response): AsyncGenerator<VersionData> {
+  // Medium-sized state machine. We swap back and forth from reading the headers <->
+  // reading data. Every chunk must contain a content-length field. If we encounter
+  // a Patches header, we count out the patches (each with headers & content) and return
+  // them inside a version.
   const enum State {
-    Headers, Data
+    VersionHeaders,
```
Comment (josephg) [inline client/client.ts:60]: Version is an ambiguous, overloaded word. `TransactionHeaders` maybe?

File: client/client.ts [hunk 10]
```diff
@@ -44,67 +45,141 @@ const searchHeaderGap = (buf: Uint8Array): null | [string, number] => {
...
 const decoder = new TextDecoder()
 
-async function *readHTTPChunks(res: Response): AsyncGenerator<PatchData> {
-  // Tiny state machine. We swap back and forth from reading the headers <->
-  // reading data. Every chunk must contain a content-length field.
+async function* readHTTPChunks(res: Response): AsyncGenerator<VersionData> {
+  // Medium-sized state machine. We swap back and forth from reading the headers <->
+  // reading data. Every chunk must contain a content-length field. If we encounter
+  // a Patches header, we count out the patches (each with headers & content) and return
+  // them inside a version.
```
Comment (josephg) [inline client/client.ts:58]: This comment is now wrong. Please update it or remove it.

File: client/client.ts [hunk 11]
```diff
@@ -44,67 +45,141 @@ const searchHeaderGap = (buf: Uint8Array): null | [string, number] => {
...
-            if (kv == null) throw Error('invalid HTTP header')
-            else return [kv[0].toLowerCase(), kv[1]]
-          }))
-
-          state = State.Data
-          buffer = buffer.slice(dataOffset)
+    while (true) {
+      // console.log('while', state, buffer.length, asciiDecoder.decode(buffer))
+      // Read as much as we can.
+      if (state == State.VersionHeaders) {
```
Comment (josephg) [inline client/client.ts:103]: switch() might be cleaner

File: client/client.ts [hunk 12]
```diff
@@ -137,49 +212,87 @@ async function *readHTTPChunks(res: Response): AsyncGenerator<PatchData> {
...
-      ...opts.reqHeaders
+      subscribe: 'keep-alive',
+      ...opts.reqHeaders,
     },
   })
 
   return {
     streamHeaders: Object.fromEntries(res.headers),
-    patches: readHTTPChunks(res)
+    versions: readHTTPChunks(res),
```
Comment (josephg) [inline client/client.ts:247]: `transactions`? `stream`?

File: client/client.ts [hunk 13]
```diff
@@ -212,28 +327,36 @@ const defaultToDoc = (contentType: string, content: Uint8Array) => {
...
-      value = (opts.toDoc ?? defaultToDoc)(contentType, data)
+      value = (opts.toDoc ?? defaultToDoc)(
+        contentType,
+        data || new Uint8Array()
+      )
       // return {value, headers}
     } else {
-      if (!opts.applyPatch) throw Error('Cannot patch documents without an apply function')
-      value = opts.applyPatch(value, patchType, data)
+      if (!opts.applyPatch)
```
Comment (josephg) [inline client/client.ts:357]: Wrap in {} if you want the body on a new line

**Review by josephg — APPROVED** [review]
Hey nice stuff with this!

I'm sort of confused with the client code now. There's 3 subscribe methods, `subscribeRaw`, `subscribe` and `subscribeApply`? The examples now still use `subscribe`, which is different from the old subscribe, but thats missing the initial version?

Whats your intent here?

A lot of my comments come down to:

- I'd like all http headers in the code to be lowercase, because they're case insensitive and that matches how http2 is used (and thus what http looks like going forward).
- I don't like the terminology of calling a transaction a "version" - since that term also refers to the version string for a document state, and its a header in transactions, and so on. I'm happy to use a different word than transaction if you have something you'd prefer, but I don't like further overloading the "version" term.

And its kind of annoying to read this diff while wading through your prettyifier tool doing lots of unrelated drive-by syntax changes. (Especially when I don't like some of them - like leaving a one-line if statement without braces). It might be a bit late now but if we're going to use a prettier I'd like to make those changes in a separate commit. And I can install the same thing too, so we don't fight over syntax. And for that, I'm curious I'm not sure if you've decided yet but I'm curious how much you're intending on working on this codebase going forward?

**Comment by canadaduane** [comment]
> Hey nice stuff with this!

Thanks for the review!

> I'm sort of confused with the client code now. There's 3 subscribe methods, `subscribeRaw`, `subscribe` and `subscribeApply`? The examples now still use `subscribe`, which is different from the old subscribe, but thats missing the initial version?
> 
> Whats your intent here?

Yeah, I wanted to chat about this, because I think my intent depends a bit on your intent :) What I'm trying to do is separate out the "pure Braid protocol" stuff from the "Braid plus possibly reasonable assumptions" aspects. For example, the previous subscribe (what is now "subscribeApply" in this PR) does some work around checking for `current-versions` and whether or not we've received versions up to that version. It also has some things around patch-id and other logic that I wanted to keep separate. Lastly, I noticed that examples 1 & 2 don't really use the capabilities of `subscribeApply` (they could quite easily use subscribeRaw), and example 3 has its own `subscribeOT`. So it made me wonder if `subscribeApply` might not be quite what we need, but I kept it around because I didn't want to assume it is not developing into the direction you want it to. What I'd like out of a clean "subscribe" is something that doesn't make assumptions, but all other subscribes could be built on. Something like your subscribeRaw, but with just a little more handling of content-type, for example.

 
> A lot of my comments come down to:
> 
> * I'd like all http headers in the code to be lowercase, because they're case insensitive and that matches how http2 is used (and thus what http looks like going forward).

Ok, no problem. I wondered about this choice. I don't have a strong preference, but I guess I have a small preference for consistency. I noticed that SOME of the headers (presumably the ones that `polka` responds with?) are in Capitalized-Kabob-Case. It's all in one commit that can be reverted.

> * I don't like the terminology of calling a transaction a "version" - since that term also refers to the version string for a document state, and its a header in transactions, and so on. I'm happy to use a different word than transaction if you have something you'd prefer, but I don't like further overloading the "version" term.

Yeah, so I think we should talk about this. I am very split about it. I think you're right. And on the other hand, I really appreciate code that uses terminology from the specification if such a spec exists. I was introducing the term "patch" here, which obviously made the existing references to "patches" ambiguous. So I switched to using the term "Version" because that's what current braid calls the concept. Transaction seems a bit heavy-handed a term to me. It's more like a "PatchBundle" or something, but that is also confusing. Maybe we can get support for changing the term in the spec?

> And its kind of annoying to read this diff while wading through your prettyifier tool doing lots of unrelated drive-by syntax changes. (Especially when I don't like some of them - like leaving a one-line if statement without braces). It might be a bit late now but if we're going to use a prettier I'd like to make those changes in a separate commit. And I can install the same thing too, so we don't fight over syntax. And for that, I'm curious I'm not sure if you've decided yet but I'm curious how much you're intending on working on this codebase going forward?

Ah, sorry, I thought you'd seen my addition previously when I added prettier rules. It can definitely be annoying. For the sake of potential redemption, I did keep it all in one commit (there are 5 separate commits in this PR). I'd be happy to reverse it if you'd like, and make it a clean change separately.

**Comment by canadaduane** [comment]
> And for that, I'm curious I'm not sure if you've decided yet but I'm curious how much you're intending on working on this codebase going forward?

I wish I could be clear here, but I think this depends on Mike as well, so I'm a bit unsure what to tell you. I very much like your code, style, etc. as it works for my brain. OTOH I'm still getting my feet wet with how he'd like to do things.

**Comment by josephg** [comment]
> What I'd like out of a clean "subscribe" is something that doesn't make assumptions, but all other subscribes could be built on. Something like your subscribeRaw, but with just a little more handling of content-type, for example.

Yeah I hear that. I'm still not entirely sure where the lines should be drawn, but I could imagine another library or something with all this logic if we wanted to design it that way.

Maybe we should move everything other than `subscribeRaw` into a separate file. The higher level APIs aren't allowed to use anything other than subscribeRaw - and for now we can consider all that code more experimental.

> Maybe we can get support for changing the term in the spec?

Yeah transaction does sound heavy. Adding a term into the spec is my instinct as well. Its a separate concept from the other concepts in the spec - and it needs a name. Might be good to drag Mike in on a conversation about this. Other ideas:

- Operation
- Change
- Change set
- Transaction

Maybe "change set" if we don't like "transaction". But I like implying transactional semantics for groups of patches. Anyway, lets pull mike in and talk about it with our voices at some point.

> Ah, sorry, I thought you'd seen my addition previously when I added prettier rules

Yeah thats fine.

**Comment by toomim** [comment]
About `transaction` vs `version`, Seph raises the point that there are multiple concepts:
 1. A bundle of patches that occur atomically at a point in time
 2. The identifier for that bundle of patches
 3. The HTTP header in which you put the identifier for the bundle of patches

In my interpretation, the Braid spec names those concepts with these terms, respectively:
 1. Version
 2. Version ID
 3. `Version:` header

I think Seph is proposing something more like:
  1. Transaction
  2. Version ID
  3. `Version:` header

My analysis favors the former. They key issue that Seph raised is distinguishing (1) from (2). I think that is easy enough to do with "Version" vs. "Version ID". This has the advantage of making clear that the two concepts are related, and only differ in that one is the ID for the other. Switching to a different word like "Transaction" brings up the possibility that these concepts could be entirely different. A reader might think that a single transaction could hold multiple versions, for instance.

There is also a question of whether the header should be named `Version:` or `Version-ID:`. One might argue that the header should be called `Version-ID:`, since it's containing an ID, and not the full data structure for a version. I sympathize with this argument, but the `Version:` header actually provides a dual role: it not only states the version ID, and also indicates the beginning of the Version data structure in the HTTP message.

In any case, it'd be nice to resolve this so that we are using the same terms. I've been finding the conversions confusing when different people use different terms for the same things.

**Comment by toomim** [comment]
I'll add that there is a more colloquial definition for "Version" as "the state of the resource at a point in time."

The state of a resource at a point of time can be referenced by its Version ID, and it can be computed by applying the bundle of patches onto its parent versions. However, I think the more fundamental concept of a version is the state of the resource itself.

**Comment by josephg** [comment]
> I'll add that there is a more colloquial definition for "Version" as "the state of the resource at a point in time."

Sure. But the set of grouped patches in a subscription *isn’t* this! The set of patches takes the document from version A to version B.

I’m hearing a desire to use the word version to mean:

- The last set of patches that are applied to bring a document to a certain version ID and
- The state of the document after that set of patches has been applied

These are distinct ideas, novel to many and adjacent in space. You might not find the overlap confusing mike but I certainly do. And if I find something confusing, I put money on other people finding it confusing as well. I think about this like UX feedback - if 1/2 expert users of the terminology find it confusing (1/3?), that might be good enough to pick another word to fix the UX. (And in this case, make the semantic difference clear through word difference.) Especially when we have lots of good words already (“change set” / “patch set” / “transaction” are my favs)

I insist on separate words here in this repo - if for no other reason than I don’t want to confuse my future self when I’m reading this code later. But it’d be nice to have consistent terms in our projects if we can.

**Comment by canadaduane** [comment]
I think it's important to note that for whatever it is that we're describing, it's orthogonal to having a `Version:` header in it. For example, a subscription stream of these things can be used to update a clock and may not contain a version.

What about the term "Update"?

- An Update can optionally have a version (ID)
- An Update doesn't imply multiple versions
- An Update can be a snapshot
- An Update can contain a bundle of patches
- Colloquially, an "update" brings you up to date on the news. "What's the latest update?"

File: examples/1/client.js (hunk 8 above)
Comment (canadaduane) [reply examples/1/client.js:5]: I'm trying to move away from an "initial value" concept in the basic subscribe function. As I understand it, a Braid subscribe is simply a stream of updates. Some of the updates (notably the first one) COULD be a snapshot, but I don't think we need that assumption in place for basic subscribe?

**Review by canadaduane — COMMENTED** [review]
I removed the Upper-Case header names, and rebased from main (incorporating the new headersToBuf).

**Comment by josephg** [comment]
Yeah I'm happy with update. Frankly anything but version. Lets talk about it next time we chat.

> I'm trying to move away from an "initial value" concept in the basic subscribe function.

You usually want an initial value when you're consuming from the API though. Like, if you're hooking it up to a svelte component, you want to render with the initial value then subscribe to subsequent values. I agree that `rawSubscribe` should just yield the series of updates - but the API most users actually want yields the series of values - starting at the first one. There's an open question of how we cut the cake here.

**Comment by josephg** [comment]
Nice one :)
